### PR TITLE
CDP Daily entity upload to weekly folder

### PR DESCRIPTION
**Pull request by binh-dam-ibigroup**
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [ ] The description lists all applicable issues this PR seeks to resolve
- [x] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

This PR contains changes to support daily upload to a per-week (Monday-Sunday) folder on AWS S3 of OTP-middleware Mongo entities.

New config parameters (see schema/README):

- `CONNECTED_DATA_PLATFORM_REPORTING_INTERVAL`: How often to aggregate and upload usage data.
- `CONNECTED_DATA_PLATFORM_FOLDER_GROUPING`: How to organize uploaded files.
- `CONNECTED_DATA_PLATFORM_REPORTED_ENTITIES` What entities to report, how, and for trip requests, whether to anonymize. No entities are reported if this is omitted.
- `CONNECTED_DATA_PLATFORM_UPLOAD_BLANK_FILES` Whether to upload files where no records have been written. Defaults to true.

Let me know if additional tests are needed for full-collection reporting.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -257,10 +257,14 @@ The special E2E client settings should be defined in `env.yml`:
...
+| CONNECTED_DATA_PLATFORM_FOLDER_AGGREGATION | string | Optional | none, weekly-monday-sunday | Specifies how to organize uploaded files. 'none' puts all uploaded files to the same folder CONNECTED_DATA_PLATFORM_S3_FOLDER_NAME. 'weekly-monday-sunday' creates a folder of the form `<CONNECTED_DATA_PLATFORM_S3_FOLDER_NAME>_<YYYY-MM-DD>_<YYYY-MM-DD>`  Defaults to none. |
 | CONNECTED_DATA_PLATFORM_S3_BUCKET_NAME | string | Optional | bucket-name | Specifies the S3 bucket name for the CDP trip history push. |
 | CONNECTED_DATA_PLATFORM_S3_FOLDER_NAME | string | Optional | folder-name | Specifies the S3 folder name for the CDP trip history push. |
+| CONNECTED_DATA_PLATFORM_REPORTED_ENTITIES | object | Optional | { "MonitoredTrip": "all", "OtpUser": "all", "TripRequest": "interval" } | Use 'all' to report all full records an entity. Use 'interval' to report full records whose dateCreated is within a reporting interval (e.g. day, hour). For TripRequest, you can use 'all anonymized' or 'interval anonymized' to anonymize records. Omitted entities are ignored. |
```
Comment (br648) [inline README.md]: Should this `Use 'all' to report all full records an entity.` be just this `Use 'all' to report all full records.`?

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/ConnectedDataManager.java [hunk 2]
```diff
@@ -182,38 +216,117 @@ private static int streamAnonymousTripsToFile(
...
+     * If no records are found, this will produce a file with an empty JSON array "[]".
+     */
+    private static int streamCollectionToFile(
+        String pathAndFileName,
+        FindIterable<?> findIterable,
+        long count
+    ) throws IOException {
+
+        long numTripRequestsWrittenToFile = 0;
+        long pos = 0;
```
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/connecteddataplatform/ConnectedDataManager.java]: Could these just ints? Are you expecting a lot of data?

File: src/main/resources/env.schema.json [hunk 3]
```diff
@@ -69,17 +84,44 @@
       "examples": ["folder-name"],
       "description": "Specifies the S3 folder name for the CDP trip history push."
     },
+    "CONNECTED_DATA_PLATFORM_REPORTED_ENTITIES": {
+      "type": "object",
+      "examples": ["{ \"MonitoredTrip\": \"all\", \"OtpUser\": \"all\", \"TripRequest\": \"interval\" }"],
+      "description": "Use 'all' to report all full records an entity. Use 'interval' to report full records whose dateCreated is within a reporting interval (e.g. day, hour). For TripRequest, you can use 'all anonymized' or 'interval anonymized' to anonymize records. Omitted entities are ignored.",
```
Comment (br648) [inline src/main/resources/env.schema.json]: Same question as in the README.md

File: README.md [hunk 4]
```diff
@@ -257,10 +257,14 @@ The special E2E client settings should be defined in `env.yml`:
 | BUGSNAG_EVENT_REQUEST_JOB_DELAY_IN_HOURS | integer | Optional | 24 | Frequency in hours to trigger event requests. |
 | BUGSNAG_PROJECT_NOTIFIER_API_KEY | string | Optional | 123e4567e89b12d3a4564266 | A valid Bugsnag project API key. |
 | BUGSNAG_REPORTING_WINDOW_IN_DAYS | integer | Optional | 14 | Specifies how far in the past events should be retrieved. |
+| BUGSNAG_WEBHOOK_PERMITTED_IPS | string | Optional | 104.196.245.109, 104.196.254.247 | Bugsnag IP addresses which webhook requests are expected to come from. |
+| CONNECTED_DATA_PLATFORM_AGGREGATION_INTERVAL | string | Optional | daily, hourly | Specifies how often to aggregate and upload usage data. Defaults to hourly. |
```
Comment (br648) [inline README.md]: Would this be better named reporting interval as referred to on line 265? Or perhaps change line 265 to aggregation interval?

**Review by br648 — CHANGES_REQUESTED** [review]
Looks good, a few minor comments. It also looks like only TripRequest - interval anonymized is tested. Would there be benefit in testing other entities with "all" and interval without anonymized?

File: README.md (hunk 4 above)
Comment (binh-dam-ibigroup) [reply README.md]: I like "reporting" better, thanks for the suggestion. I also renamed "folder aggregation" into "folder grouping" (https://github.com/ibi-group/otp-middleware/pull/254/commits/48e99e954615e61b0d9de1f581ee701eb08b5300).

**Review by br648 — APPROVED** [review]
Has the test data been updated to work with OTP2? I wasn't able to run the unit tests E2E (region related error), I'm assuming you have tested these so approving.

**Comment by binh-dam-ibigroup** [comment]
> Has the test data been updated to work with OTP2? 

For test data, only `TripRequest` follows OTP2.  `TripSummary` and `MonitoredTrip` are awaiting #243. `OtpUser` and `TrackedJourney` are OTP-agnostic.

> I wasn't able to run the unit tests E2E (region related error), I'm assuming you have tested these

The updated snapshots are the result of that.

File: src/main/java/org/opentripplanner/middleware/connecteddataplatform/ConnectedDataManager.java [hunk 5]
```diff
@@ -45,9 +51,11 @@
  */
 public class ConnectedDataManager {
 
-    public static final String FILE_NAME_SUFFIX = "anon-trip-data";
-    public static final String ZIP_FILE_NAME_SUFFIX = FILE_NAME_SUFFIX + ".zip";
-    public static final String DATA_FILE_NAME_SUFFIX = FILE_NAME_SUFFIX + ".json";
+    public static final String ANON_TRIP_FILE_NAME = "anon-trip-data";
+    public static final String ANON_TRIP_ZIP_FILE_NAME = ANON_TRIP_FILE_NAME + ".zip";
+    public static final String ANON_TRIP_JSON_FILE_NAME = ANON_TRIP_FILE_NAME + ".json";
```
Comment (miles-grant-ibigroup) [inline src/main/java/org/opentripplanner/middleware/connecteddataplatform/ConnectedDataManager.java:56]: If we have `ZIP_FILE_EXTENSION` shouldn't we use it here?

**Review by miles-grant-ibigroup — APPROVED** [review]
Can't comment too much on larger architecture decisions here but the code looks clean and makes sense.

**Pull request merged**